File: 2.两数相加.js
/*
 * @lc app=leetcode.cn id=2 lang=javascript
 *
 * [2] 两数相加
 */
/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var addTwoNumbers = function(l1, l2) {
    let resNode = new ListNode(0);
    let head = resNode;
    let ik = 0;

    while (l1 || l2) {
        let sum = l1.val + l2.val + ik;
        if (sum >= 10) {
            ik = 1;
            sum = sum % 10;
        } else {
            ik = 0;
        }
        resNode.val = sum;

        if (l2.next && !l1.next) {
            l1.next = new ListNode(0);
        } else if (l1.next && !l2.next) {
            l2.next = new ListNode(0);
        } else if (!l2.next && !l1.next) {
            if (ik === 1) {
                resNode.next = new ListNode(1);
            }
            break;
        }

        l1 = l1.next;
        l2 = l2.next;
        resNode.next = function() {};
        resNode = resNode.next;
    }

    return head;
};

File: 2.两数相加.py
#
# @lc app=leetcode.cn id=2 lang=python3
#
# [2] 两数相加
#

# @lc code=start
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: ListNode, l2: ListNode) -> ListNode:
        p1 = l1
        p2 = l2
        h = ListNode(None)
        ans = h
        carry = 0
        while p1 or p2:
            val = carry
            if p1:
                val += p1.val
            if p2:
                val += p2.val
            carry = val // 10
            h.next = ListNode(val % 10)
            if p1:
                p1 = p1.next
            if p2:
                p2 = p2.next
            h = h.next
        if carry:
            h.next = ListNode(carry)
        return ans.next
# @lc code=end
